[12.0][IMP] sales_team_security: Auto-followers (create and remove) in child partners

victoralmau opened the pull request:
Auto-followers (create and remove) in child partners when salesman is changed (write function)

Please @pedrobaeza review it.

Locked by: 
- [x] https://github.com/OCA/sale-workflow/issues/1446 `sale_order_lot_selection`

@Tecnativa TT27986

victoralmau committed:
[IMP] sales_team_security: Auto-followers on main & invoice/shipping contacts + others

- For avoiding access errors on this records that are related to sales flow
- Sync salesman on parent change

Co-Authored-By: Pedro M. Baeza <pedro.baeza@tecnativa.com>

victoralmau commented:
Now exists some incompatibility with `partner_contact_sale_info_propagation` addon in some tests case.
I will change to prevent it but IMO it's need to define incompatibility addon in manifest.
What do you think @pedrobaeza ?

rousseldenis commented:
@victoralmau could you add a little explanation of this addition in DESCRIPTION.rst ?

pedrobaeza commented on sales_team_security/tests/test_sales_team_security.py:
```diff
@@ -34,6 +51,17 @@ def setUpClass(cls):
                 (4, cls.env.ref("sales_team.group_sale_salesman").id)],
             "sale_team_id": cls.team.id,
         })
+        cls.partner_contact_sale_info_propagation_installed = cls._is_module_installed(
```
You should make tests independently from this module. If it's interfering, then the problem in in the other module, and would require a mitigation method like this one:

https://github.com/OCA/bank-payment/blob/3c96fedbb95060c9cd3a2aa87d383324d486bee1/account_payment_transfer_reconcile_batch/models/payment_order.py#L28

victoralmau replied:
I understand what you say and your proposal, but I don't really know how to interfere with it.
The functionality that is added in this PR auto-add followers of the salesmans (`user_id`) of the contacts. (only in company contacts). The tests check this and work fine.
The `partner_contact_sale_info_propagation` addon auto-define salesman (`user_id`) for all parent contacts, which causes the expected result of the tests to change. Any proposal?

pedrobaeza replied:
Yes, to avoid such propagation on test mode if you are not testing the other module, which style is the link I sent you.

pedrobaeza commented on sales_team_security/migrations/12.0.3.0.0/post-migration.py:
```diff
@@ -0,0 +1,17 @@
+@openupgrade.migrate()
+def migrate(env, version):
+    """At installation time, need to followers sync"""
+    partners = env['res.partner'].search([
```
This should be a post_init_hook, but also being called from this migration script.

pedrobaeza commented on sales_team_security/models/res_partner.py:
```diff
@@ -40,3 +41,54 @@ def _onchange_parent_id_sales_team_security(self):
         if self.parent_id and self.parent_id.team_id and not self.team_id:
             self.team_id = self.parent_id.team_id.id
+
+    @api.depends('company_type', 'child_ids', 'child_ids.user_id')
```
Depends is only for computed field, so this doesn't add anything.

victoralmau replied:
Changed it to `commercial_partner_id`

pedrobaeza commented on sales_team_security/models/res_partner.py:
```diff
@@ -40,3 +41,54 @@ def _onchange_parent_id_sales_team_security(self):
+                    partner_ids=user_ids.mapped('partner_id').ids
+                )
+
+    @api.depends('parent_id', 'child_ids')
```
Remove it

pedrobaeza commented on sales_team_security/models/res_partner.py:
```diff
@@ -40,3 +41,54 @@ def _onchange_parent_id_sales_team_security(self):
+            if partner:
+                partner.sync_followers_in_childs()
+
+    @api.depends('company_type', 'parent_id', 'child_ids')
```
Remove it

pedrobaeza commented on sales_team_security/models/res_partner.py:
```diff
@@ -40,3 +41,54 @@ def _onchange_parent_id_sales_team_security(self):
+                partner.sync_followers_in_childs()
+
+    @api.depends('company_type', 'parent_id', 'child_ids')
+    def get_partner_to_sync(self):
```
Why not using `self.commercial_partner_id`?

pedrobaeza commented on sales_team_security/models/res_partner.py:
```diff
@@ -40,3 +41,54 @@ def _onchange_parent_id_sales_team_security(self):
+                if partner_to_sync:
+                    ids = (partner_to_sync.child_ids.ids + partner_to_sync.ids)
+                    partners = self.env['res.partner'].search([('id', 'in', ids)])
+                    partners.message_unsubscribe(
```
Why this code is not in `sync` method? If you call it sync, do the real sync, not only half of the task.

pedrobaeza commented on sales_team_security/models/res_partner.py:
```diff
@@ -40,3 +41,54 @@ def _onchange_parent_id_sales_team_security(self):
+    @api.depends('company_type', 'child_ids', 'child_ids.user_id')
+    def sync_followers_in_childs(self):
+        for item in self.filtered(lambda x: x.company_type == 'company'):
+            partners = self.env['res.partner'].search([
```
I don't see here any extraction of the delivery / invoicing contacts. There's a method for that in the partner.

victoralmau replied:
Function apply to `child` ids

pedrobaeza reviewed (approved):
Adjusted for syncing followers according invoice and delivery address + main contact.

pedrobaeza committed:
[FIX] sales_team_security: Adjust security rules for team managers

A team manager wasn't able to access to their orders if the team was
different from their team.

Whole review has been done for the rest of the channel manager rules,
and added some tests.

Migration script included.

victoralmau commented:
Actually, test fails when `partner_contact_sale_info_propagation` installed. 
I'm not really sure if both addons can work fine together and maybe solution can be add incompatibility in manifest file.
What do you think about it?

victoralmau commented:
It's ready to merge now, right?

pedrobaeza commented:
/ocabot merge nobump

pedrobaeza commented:
@victoralmau Travis is red

victoralmau commented:
> @victoralmau Travis is red

I have added a commit to resolve Travis. It is related to https://github.com/OCA/sale-workflow/pull/1439#issuecomment-866634495. (This will be necessary in v13 as well).

pedrobaeza commented:
The patch for making this green is not correct, as you should instead prevent in the other module to propagate the information if you are on test mode and not testing your module (through the context key).

victoralmau committed:
[IMP] partner_contact_sale_info_propagation: Add context to prevent change partner propagation.
TT27986

victoralmau commented:
Thanks Pedro, would the current change be correct?

pedrobaeza commented on partner_contact_sale_info_propagation/models/res_partner.py:
```diff
@@ -1,50 +1,64 @@
+            not config["test_enable"]
+            or (
+                config["test_enable"]
+                and not self.env.context.get("prevent_propagation", False)
```
This should be the exact contrary:
```suggestion
                and not self.env.context.get("test_propagation")
```
and only enable such context on this module tests.

pedrobaeza commented on partner_contact_sale_info_propagation/models/res_partner.py:
```diff
@@ -1,50 +1,64 @@
-                if childs:
-                    childs.write({'team_id': vals['team_id']})
-        return super(ResPartner, self).write(vals)
+        if self._check_propagation_allowed():
```
Avoid this diff doing an early return:
```suggestion
        if not self._check_propagation_allowed():
            return super().write(vals)
```

pedrobaeza commented on partner_contact_sale_info_propagation/models/res_partner.py:
```diff
@@ -1,50 +1,64 @@
-            if not self.team_id:
-                res.setdefault('value', {}).update(team_id=parent.team_id)
+        res = super().onchange_parent_id()
+        if self._check_propagation_allowed():
```
Is this also needed?

victoralmau replied:
I have verified that this is not necessary.

pedrobaeza commented on partner_contact_sale_info_propagation/models/res_partner.py:
```diff
@@ -1,44 +1,57 @@
-                if childs:
-                    childs.write({'team_id': vals['team_id']})
-        return super(ResPartner, self).write(vals)
+        if not self._check_propagation_allowed():
```
You are still putting a big diff as not doing the early return.

victoralmau replied:
Sorry, now the diff is smaller.

pedrobaeza commented:
/ocabot merge nobump
